### ISSUE TITLE
Collection inheritance

### ISSUE THREAD

**Issue by henri-hulski** [ponyorm/pony#212]
I want to create a security system let's say similar to the one ponyJS uses but database driven. 

If I have the following entities:
```python
class Principal(db.Entity):
    name = Required(str)
    groups = Set('Group')

class User(Principal):
    email = Required(str) 

class Group(Principal):
    members = Set("Principal")
```
Now I have the groups 'Admin', which is member of the group 'Moderator', which is member of the group 'Editor'. So that every Admin is also Moderator and Editor and every Moderator is also Editor. 

I want to check if a User is an Editor. The User 'Leader' is only in the Admin group.

Is it possible to get him with 
```python
>>> u = User(name='Leader')
>>> g = Group(name='Editor')

>>> u in g.members
True
```

In other words have collections inheritance?

**Comment by henri-hulski** [comment]
I just tried it out:

```python
>>> leader = User(name='Leader', email='leader@post.com')

>>> admin = Group(name='Admin')
>>> moderator = Group(name='Moderator')
>>> editor = Group(name='Editor')

>>> moderator.members.add(admin)
>>> editor.members.add(moderator)
>>> admin.members.add(leader)

>>> leader in editor.members
False
>>> leader in admin.members
True
```
So it seems not to work.
Is there a way to achieve this?

**Comment by blackshirt** [comment]
is this similar problem i'm trying to grasp ?
here the stackoverflow.com link i pasted, http://stackoverflow.com/questions/41030262/pony-orm-permissionerror-in-access-rule-for-to-json-function

**Comment by henri-hulski** [comment]
I don't use PonyJS. I only mentioned it because I try to set up a similar security model. But my model is database driven so for now I have a User and Group model which inherit from the Principal model. 
I want to have a Group inheritance so that when the Admin group is member of the Moderator group every user who is in the Admin group should also have Moderator permissions. 
Now when I test if a User has Moderator permission this should also be true for a User from the Admin group.

I want to use this together with the [Morepath security infrastructure](http://morepath.readthedocs.io/en/latest/security.html).

BTW PonyJS seems unfinished and [abandoned]( https://disqus.com/embed/comments/?base=default&version=e8120450c59f0ede4efdd2c4258714db&f=ponyorm&t_u=https%3A%2F%2Fdocs.ponyorm.com%2Fponyjs.html&t_d=PonyJS%20-%20using%20Pony%20ORM%20objects%20at%20the%20frontend¶&t_t=PonyJS%20-%20using%20Pony%20ORM%20objects%20at%20the%20frontend¶&s_o=default#).

**Comment by kozlovsky** [comment]
I'd suggest to have an additional collection attribute which contains explicitly stated transitive relationships between groups:

```python
class Principal(db.Entity):
    name = Required(str)

class Group(Principal):
    direct_bases = Set('Group', revert='direct_subgroups')
    direct_subgroups = Set('Group')
    all_bases = Set('Group', revert='all_subgroups')  # contains transitive relation
    all_subgroups = Set('Group')
    users = Set(User)

class User(Principal):
    groups = Set('Group')
```

Now when creating a new group we can specify all bases on which this group is depends on:

```python
editor = Group(name='editor')
moderator = Group(name='moderator', direct_bases=[editor], all_bases=[editor])
admin = Group(name='admin', direct_bases=[moderator], all_bases=[moderator, editor])
```

Also we need to add group itself to its `all_bases` collection:

```python
editor.all_bases.add(editor)
moderator.all_bases.add(moderator)
admin.all_bases.add(admin)
```

Now let's create a new admin user:
```python
u = User(name='Leader', groups=[admin])
```

In order to check if a user is belongs to a specific group, we can use the following expression:
```python
editor in u.groups.all_bases
```
or, equivalently
```python
u in editor.all_subgroups.users
```

These expressions use Pony attribute lifting expression syntax. `u.groups` is a collection of `Group` objects, and in Pony each collection has all attributes of its members, and the value of each attribute is a union of all values of the same attribute for individual collection items. So `u.groups.all_bases` is a union of all bases for all groups of a specific user. The same is for `editor.all_subgroups.users`.

**Comment by henri-hulski** [comment]
@kozlovsky Thanks!
This looks really interesting. 
I think I can also remove the Principal entity as it was mainly to allow Users and Groups to be members of Groups. 
And I think in this case we don't need direct_bases / direct_subgroups. Maybe in other cases of inheritance it would be needed. 

I will try it out.

**Comment by kozlovsky** [comment]
It is not necessary to have direct bases / subgroups, but it can be useful to recreate transitive base / subgroups if you perform some incremental modifications of inheritance graph, for example if you delete some group

**Comment by henri-hulski** [comment]
Yeah this is why I say in other cases of inheritance it makes sense. 

But here we get more flexibility without as a group can have different base groups which not necessarily have to be in a chain.

**Comment by henri-hulski** [comment]
@kozlovsky Thanks for your help!
This works great.
